File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    int solve(vector<int>& potions, long long success, int addon, int s, int e){
        if(s > e) return 0;
        if(s == e){
            if(1LL * potions[s] * addon < success) return potions.size();
            return s;
        }

        int mid = s + (e-s)/2;
        if(1LL * potions[mid] * addon >= success){
            e = mid;
        }
        else{
            s = mid+1;
        }
        return solve(potions, success, addon, s, e);
    }
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        vector<int> ans;
        sort(potions.begin(), potions.end());
        int sz = spells.size(), sz2 = potions.size();
        for(int i=0; i<sz; i++){
            int cnt = solve(potions, success, spells[i], 0, sz2-1);
            if(cnt == sz2) cnt = 0;
            else cnt = sz2 - cnt;
            ans.push_back(cnt);
        }

        return ans;
    }
};

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length, m = potions.length;
        int ans[] = new int[n];
        Arrays.sort(potions);

        for(int i=0; i<n; i++){
            ans[i] = m - solve(potions, m, spells[i], success);
        }

        return ans;
    }

    private int solve(int[] potions, int m, long src, long dst){
        int s = 0, e = m-1;

        while(s <= e){
            int mid = s + (e-s)/2;
            // System.out.println(s + " " + e + " " + mid + " " + src);
            if(potions[mid] * src >= dst){
                e = mid - 1;
            }
            else {
                s = mid + 1;
            }
        }

        // System.out.println(s + " " + e);

        return s;
    }
}